File: app/src/org/commcare/fragments/connect/ConnectJobIntroFragment.java
package org.commcare.fragments.connect;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavDirections;
import androidx.navigation.fragment.NavHostFragment;

import org.commcare.AppUtils;
import org.commcare.android.database.connect.models.ConnectJobRecord;
import org.commcare.android.database.connect.models.ConnectLearnModuleSummaryRecord;
import org.commcare.android.database.connect.models.ConnectUserRecord;
import org.commcare.connect.ConnectAppLaunchController;
import org.commcare.connect.ConnectDateUtils;
import org.commcare.connect.database.ConnectJobUtils;
import org.commcare.connect.database.ConnectUserDatabaseUtil;
import org.commcare.connect.network.connect.ConnectApiHandler;
import org.commcare.connect.network.PersonalIdOrConnectApiErrorHandler;
import org.commcare.dalvik.R;
import org.commcare.dalvik.databinding.FragmentConnectJobIntroBinding;
import org.commcare.google.services.analytics.FirebaseAnalyticsUtil;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * Fragment for showing detailed info about an available job
 *
 * @author dviggiano
 */
public class ConnectJobIntroFragment extends ConnectJobFragment<FragmentConnectJobIntroBinding> {

    public ConnectJobIntroFragment() {
        // Required empty public constructor
    }

    @Override
    public @NotNull View onCreateView(@NotNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        requireActivity().setTitle(R.string.connect_job_info_view_opportunity);

        ActionBar actionBar = ((AppCompatActivity)requireActivity()).getSupportActionBar();
        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(R.drawable.ic_connect_close);
        }

        getBinding().tvJobTitle.setText(job.getTitle());
        getBinding().tvJobDescription.setText(job.getDescription());
        getBinding().tvExpiryValue.setText(ConnectDateUtils.formatShortDate(job.getProjectEndDate()));
        getBinding().tvMaxEarningsValue.setText(job.getMoneyStringWithSymbol(job.getTotalBudget()));

        getBinding().btnStart.setOnClickListener(v -> startLearning());

        populateLearnCard();
        populateDeliveryCards();

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ActionBar actionBar = ((AppCompatActivity)requireActivity()).getSupportActionBar();
        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(0);
        }
    }

    private void populateLearnCard() {
        List<ConnectLearnModuleSummaryRecord> modules = job.getLearnAppInfo().getLearnModules();
        int totalHours = 0;
        for (ConnectLearnModuleSummaryRecord module : modules) {
            totalHours += module.getTimeEstimate();
        }

        getBinding().cardLearnModules.setValueText(String.valueOf(modules.size()));
        getBinding().cardLearnModules.setSubtitleText(
                getString(R.string.connect_opportunity_learn_hours_total, totalHours));
        getBinding().cardLearnModules.setOnCardClick(() -> {
            NavHostFragment.findNavController(this).navigate(
                    ConnectJobIntroFragmentDirections
                            .actionConnectJobIntroFragmentToConnectLearnModulesBottomSheet());
            return null;
        });
    }

    private void populateDeliveryCards() {
        getBinding().cardMaxVisits.setValueText(String.valueOf(job.getMaxPossibleVisits()));
        getBinding().cardMaxVisits.setSubtitleText(
                getString(R.string.connect_opportunity_visits_per_day, job.getMaxDailyVisits()));

        getBinding().cardDays.setValueText(String.valueOf(job.getDaysRemaining()));

        getBinding().cardMaxEarnings.setValueText(job.getMoneyStringWithSymbol(job.getTotalBudget()));
        getBinding().cardMaxEarnings.setSubtitleText(
                getString(R.string.connect_opportunity_payment_units, job.getPaymentUnits().size()));
    }

    private void startLearning() {
        ConnectUserRecord user = ConnectUserDatabaseUtil.getUser(getContext());

        new ConnectApiHandler<Boolean>() {
            @Override
            public void onFailure(@NonNull PersonalIdOrConnectApiErrorCodes errorCode, @Nullable Throwable t) {
                reportApiCall(false);
                if (!isAdded()) {
                    return;
                }

                String error = PersonalIdOrConnectApiErrorHandler.handle(requireActivity(), errorCode, t);
                if (PersonalIdOrConnectApiErrorHandler.isNetworkError(errorCode)) {
                    showError(getString(R.string.failed_to_start_learning));
                } else {
                    navigateToMessageDisplayDialog(
                            getString(R.string.error),
                            error,
                            false,
                            R.string.ok);
                }
            }

            @Override
            public void onSuccess(Boolean success) {
                hideError();
                reportApiCall(success);

                job.setStatus(ConnectJobRecord.STATUS_LEARNING);
                ConnectJobUtils.upsertJob(getContext(), job);

                if (!isAdded()) {
                    return;
                }

                String appId = job.getLearnAppInfo().getAppId();
                boolean appInstalled = AppUtils.isAppInstalled(appId);
                if (appInstalled) {
                    new ConnectAppLaunchController(ConnectJobIntroFragment.this)
                            .launchApp(appId, true, ConnectJobIntroFragment.this::popSelfOnceHidden);
                } else {
                    String title = getString(R.string.connect_downloading_learn);
                    NavHostFragment.findNavController(ConnectJobIntroFragment.this).navigate(
                            ConnectJobIntroFragmentDirections.
                                    actionConnectJobIntroFragmentToConnectDownloadingFragment(
                                            title, true));
                }
            }

        }.connectStartLearning(requireContext(), user, job.getJobUUID());
    }

    private void reportApiCall(boolean success) {
        FirebaseAnalyticsUtil.reportCccApiStartLearning(success);
    }

    @Override
    protected @NotNull FragmentConnectJobIntroBinding inflateBinding(@NotNull LayoutInflater inflater, @Nullable ViewGroup container) {
        return FragmentConnectJobIntroBinding.inflate(inflater, container, false);
    }

    private void navigateToMessageDisplayDialog(@Nullable String title, @Nullable String message, boolean isCancellable, int buttonText) {
        NavDirections navDirections = ConnectJobIntroFragmentDirections.actionConnectJobIntroFragmentToPersonalidMessageDisplayDialog(
                title, message, getString(buttonText), null).setIsCancellable(isCancellable);
        NavHostFragment.findNavController(this).navigate(navDirections);
    }
}

File: app/src/org/commcare/fragments/connect/ConnectJobIntroFragment.kt
package org.commcare.fragments.connect

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment
import org.commcare.android.database.connect.models.ConnectJobRecord
import org.commcare.connect.ConnectDateUtils
import org.commcare.connect.ConnectMoneyUtils
import org.commcare.connect.database.ConnectJobUtils
import org.commcare.connect.network.base.PersonalIdOrConnectApiErrorHandler
import org.commcare.connect.repository.DataState
import org.commcare.connect.viewmodel.ConnectJobIntroViewModel
import org.commcare.connect.viewmodel.InstallState
import org.commcare.dalvik.R
import org.commcare.dalvik.databinding.FragmentConnectJobIntroBinding
import org.commcare.fragments.extensions.hasLiveView
import org.commcare.google.services.analytics.FirebaseAnalyticsUtil
import org.commcare.views.connect.renderInstallState
import java.text.DateFormat

/**
 * Fragment for showing detailed info about an available job
 *
 */
class ConnectJobIntroFragment : ConnectJobFragment<FragmentConnectJobIntroBinding>() {
    private lateinit var viewModel: ConnectJobIntroViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        requireActivity().setTitle(R.string.connect_job_info_view_opportunity)

        (requireActivity() as AppCompatActivity)
            .supportActionBar!!
            .setHomeAsUpIndicator(R.drawable.ic_connect_close)

        viewModel =
            ViewModelProvider(
                this,
                ViewModelProvider.AndroidViewModelFactory.getInstance(requireActivity().application),
            )[ConnectJobIntroViewModel::class.java]

        observeStartLearning()

        binding.tvJobTitle.text = job.title
        binding.tvJobDescription.text = job.shortDescription
        binding.tvExpiryValue.text =
            ConnectDateUtils.formatDate(job.projectEndDate, DateFormat.SHORT)
        binding.tvMaxEarningsValue.text =
            ConnectMoneyUtils.moneyStringWithSymbol(job.currency, job.totalBudget)

        binding.connectIntroCtaBar.setOnCtaClickListener { startLearning() }

        populateLearnCard()
        populateDeliveryCards()

        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        (requireActivity() as AppCompatActivity).supportActionBar!!.setHomeAsUpIndicator(0)
    }

    private fun observeStartLearning() {
        viewModel.startLearning.observe(viewLifecycleOwner) { state ->
            when (state) {
                is DataState.Loading -> {
                    showLoading()
                }

                is DataState.Success -> {
                    hideLoading()
                    reportApiCall(true)

                    job.status = ConnectJobRecord.STATUS_LEARNING
                    ConnectJobUtils.upsertJob(job)

                    if (!hasLiveView()) return@observe
                    hideError()

                    launchApp(true)
                }

                is DataState.Error -> {
                    reportApiCall(false)
                    if (!hasLiveView()) return@observe
                    hideLoading()
                    val error =
                        PersonalIdOrConnectApiErrorHandler.handle(
                            requireActivity(),
                            state.errorCode,
                            state.throwable,
                        )
                    if (PersonalIdOrConnectApiErrorHandler.isNetworkError(state.errorCode)) {
                        showError(getString(R.string.failed_to_start_learning))
                    } else {
                        navigateToMessageDisplayDialog(
                            getString(R.string.error),
                            error,
                            false,
                            R.string.ok,
                        )
                    }
                }

                else -> {}
            }
        }
    }

    private fun populateLearnCard() {
        val modules = job.learnAppInfo.learnModules
        val totalHours = modules.sumOf { it.timeEstimate }

        binding.cardLearnModules.valueText = modules.size.toString()
        binding.cardLearnModules.titleText =
            resources.getQuantityString(
                R.plurals.connect_opportunity_learn_modules_label,
                modules.size,
            )
        binding.cardLearnModules.subtitleText =
            resources.getQuantityString(
                R.plurals.connect_opportunity_estimated_hours,
                totalHours,
                totalHours,
            )
        binding.cardLearnModules.onCardClick = {
            NavHostFragment.findNavController(this).navigate(
                ConnectJobIntroFragmentDirections
                    .actionConnectJobIntroFragmentToConnectLearnModulesBottomSheet(),
            )
        }
    }

    private fun populateDeliveryCards() {
        binding.cardMaxVisits.valueText = job.maxPossibleVisits.toString()
        binding.cardMaxVisits.subtitleText =
            getString(R.string.connect_opportunity_visits_per_day, job.maxDailyVisits)

        binding.cardDays.valueText = job.daysRemaining.toString()

        binding.cardMaxEarnings.valueText =
            ConnectMoneyUtils.moneyStringWithSymbol(job.currency, job.totalBudget)
        binding.cardMaxEarnings.subtitleText =
            resources.getQuantityString(
                R.plurals.connect_opportunity_payment_units,
                job.paymentUnits.size,
                job.paymentUnits.size,
            )
    }

    private fun startLearning() {
        viewModel.startLearning(job.jobUUID)
    }

    /** The learn app downloads into the same bar the user pressed to start learning. */
    override fun onInstallStateChanged(
        state: InstallState?,
        isLearning: Boolean,
    ) = binding.connectIntroCtaBar.renderInstallState(state, isLearning, ::forgetInstallFailure)

    private fun reportApiCall(success: Boolean) {
        FirebaseAnalyticsUtil.reportCccApiStartLearning(success)
    }

    override fun inflateBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
    ): FragmentConnectJobIntroBinding = FragmentConnectJobIntroBinding.inflate(inflater, container, false)

    private fun navigateToMessageDisplayDialog(
        title: String,
        message: String,
        isCancellable: Boolean,
        buttonText: Int,
    ) {
        val navDirections =
            ConnectJobIntroFragmentDirections
                .actionConnectJobIntroFragmentToPersonalidMessageDisplayDialog(
                    title,
                    message,
                    getString(buttonText),
                    null,
                ).setIsCancellable(isCancellable)
        NavHostFragment.findNavController(this).navigate(navDirections)
    }
}
